selene-lib/src/lints/roblox_complex_udim2_new.rs

```diff
@@ -0,0 +1,176 @@
+use super::*;
+use crate::ast_util::range;
+use std::convert::Infallible;
+
+use full_moon::{
+    ast::{self, Ast},
+    visitors::Visitor,
+};
+
+pub struct ComplexUDim2NewLint;
+
+fn create_diagnostic(mismatch: &UDim2ComplexArgs) -> Diagnostic {
+    let code = "roblox_complex_udim2_new";
+    let primary_label = Label::new(mismatch.call_range);
+
+    if mismatch.no_scale {
+        Diagnostic::new_complete(
+            code,
+            "this UDim2.new call only sets offset, and can be simplified using UDim2.fromOffset"
+                .to_owned(),
+            primary_label,
+            vec![format!(
+                "try: UDim2.fromOffset({}, {})",
+                mismatch.arg_0, mismatch.arg_1
+            )],
+            Vec::new(),
+        )
+    } else {
+        Diagnostic::new_complete(
+            code,
+            "this UDim2.new call only sets scale, and can be simplified using UDim2.fromScale"
+                .to_owned(),
+            primary_label,
+            vec![format!(
+                "try: UDim2.fromScale({}, {})",
+                mismatch.arg_0, mismatch.arg_1
+            )],
+            Vec::new(),
+        )
+    }
+}
+
+impl Lint for ComplexUDim2NewLint {
+    type Config = ();
+    type Error = Infallible;
+
+    const SEVERITY: Severity = Severity::Warning;
+    const LINT_TYPE: LintType = LintType::Style;
+
+    fn new(_: Self::Config) -> Result<Self, Self::Error> {
+        Ok(ComplexUDim2NewLint)
+    }
+
+    fn pass(&self, ast: &Ast, context: &Context, _: &AstContext) -> Vec<Diagnostic> {
+        if !context.is_roblox() {
+            return Vec::new();
+        }
+
+        let mut visitor = UDim2NewVisitor::default();
+
+        visitor.visit_ast(ast);
+
+        visitor.args.iter().map(create_diagnostic).collect()
+    }
+}
+
+#[derive(Default)]
+struct UDim2NewVisitor {
+    args: Vec<UDim2ComplexArgs>,
+}
+
+struct UDim2ComplexArgs {
+    call_range: (usize, usize),
+    arg_0: f32,
+    arg_1: f32,
+    no_scale: bool,
```

Comment:
Can we use an enum for the scale/offset flag? The no_scale -> suggest offset logic takes me a bit to think.

Reply:
Good call